cpp/src/structure/select_random_vertices_impl.hpp

```diff
@@ -147,8 +212,34 @@ rmm::device_uvector<vertex_t> select_random_vertices(
   }
 
   if constexpr (multi_gpu) {
+    auto vertex_partition_range_lasts =
+      given_set ? cugraph::partition_manager::compute_vertex_partition_range_lasts(
+                    handle, static_cast<vertex_t>((*given_set).size()))
+                : graph_view.vertex_partition_range_lasts();
+
     mg_sample_buffer = cugraph::detail::shuffle_int_vertices_to_local_gpu_by_vertex_partitioning(
-      handle, std::move(mg_sample_buffer), graph_view.vertex_partition_range_lasts());
+      handle, std::move(mg_sample_buffer), vertex_partition_range_lasts);
+
+    if (given_set) {
+      auto const comm_rank = handle.get_comms().get_rank();
+      auto const comm_size = handle.get_comms().get_size();
+
+      auto start_idx = comm_rank ? vertex_partition_range_lasts[comm_rank - 1] : 0;
+
+      thrust::transform(handle.get_thrust_policy(),
+                        mg_sample_buffer.begin(),
+                        mg_sample_buffer.end(),
+                        mg_sample_buffer.begin(),
+                        [start_idx = start_idx] __device__(auto idx) { return idx - start_idx; });
+    }
```

Comment:
Better use thrust::transform_iterator with https://github.com/rapidsai/cugraph/blob/branch-23.06/cpp/include/cugraph/utilities/device_functors.cuh#L113

This reduces the amount of memory traffic.